### ISSUE TITLE
Compiling error - thread.h

### ISSUE THREAD

**Issue by triangle-stream** [jprjr/icecast-hls#4]
I was trying to compile this beautiful project in order to test some functionalities but unfortunately I encountered several issues during the process.

All of the listed dependencies are installed (-dev) but the same issue is still there also if I disable the features.

Commmand:
`make ENABLE_CURL=0 ENABLE_AVFORMAT=0 ENABLE_AVCODEC=0 ENABLE_EXHALE=0 ENABLE_FDK_AAC=0 ENABLE_AVFILTER=0`

Result:
```
cc -o icecast-hls src/ini.o src/map.o src/main.o src/membuf.o src/miniflac.o src/minifmp4.o src/codecs.c src/decoder.o src/decoder_plugin.o src/decoder_plugin_auto.o src/decoder_plugin_miniflac.o src/decoder_plugin_passthrough.o src/demuxer.o src/demuxer_plugin.o src/demuxer_plugin_auto.o src/demuxer_plugin_flac.o src/demuxer_plugin_ogg.o src/destination.o src/destinationlist.o src/destination_sync.o src/encoder.o src/encoder_plugin.o src/encoder_plugin_passthrough.o src/filter.o src/filter_plugin.o src/filter_plugin_passthrough.o src/frame.o src/hls.o src/ich_time.o src/id3.o src/input.o src/input_plugin.o src/input_plugin_file.o src/input_plugin_stdin.o src/logger.o src/muxer.o src/muxer_plugin.o src/muxer_plugin_adts.o src/muxer_plugin_flac.o src/muxer_plugin_fmp4.o src/muxer_plugin_ogg.o src/muxer_plugin_ogg_flac.o src/muxer_plugin_ogg_opus.o src/muxer_plugin_packedaudio.o src/muxer_plugin_passthrough.o src/output.o src/output_plugin.o src/output_plugin_file.o src/output_plugin_icecast.o src/output_plugin_stdout.o src/output_plugin_folder.o src/packet.o src/samplefmt.o src/segment.o src/socket.o src/source.o src/sourcelist.o src/source_sync.o src/str.o src/strbuf.o src/tag.o src/tagmap.o src/tagmap_default.o src/thread.o src/version.o src/encoder_plugin_opus.o  -L/usr/local/lib -lopus
/usr/bin/ld: src/thread.o: in function `thread_create':
/user/dev/icecast-hls/src/thread.h:710: undefined reference to `pthread_create'
/usr/bin/ld: src/thread.o: in function `thread_destroy':
/user/dev/icecast-hls/src/thread.h:730: undefined reference to `pthread_join'
/usr/bin/ld: src/thread.o: in function `thread_join':
/user/dev/icecast-hls/src/thread.h:750: undefined reference to `pthread_join'
/usr/bin/ld: src/thread.o: in function `thread_detach':
/user/dev/icecast-hls/src/thread.h:767: undefined reference to `pthread_detach'
/usr/bin/ld: src/thread.o: in function `thread_tls_create':
/user/dev/icecast-hls/src/thread.h:1459: undefined reference to `pthread_key_create'
/usr/bin/ld: src/thread.o: in function `thread_tls_destroy':
/user/dev/icecast-hls/src/thread.h:1478: undefined reference to `pthread_key_delete'
/usr/bin/ld: src/thread.o: in function `thread_tls_set':
/user/dev/icecast-hls/src/thread.h:1494: undefined reference to `pthread_setspecific'
/usr/bin/ld: src/thread.o: in function `thread_tls_get':
/user/dev/icecast-hls/src/thread.h:1510: undefined reference to `pthread_getspecific'
collect2: error: ld returned 1 exit status
make: *** [Makefile:275: icecast-hls] Error 1
```
Is there anything I'm missing? If I disable only the Exhale feature (since I don't need it now), another error is displayed:
```
cc -Wall -Wextra -g -O2 -fPIC -I/usr/include/x86_64-linux-gnu -c -o src/avcodec_utils.o src/avcodec_utils.c
cc -Wall -Wextra -g -O2 -fPIC -I/usr/include/x86_64-linux-gnu -c -o src/avframe_utils.o src/avframe_utils.c
cc -Wall -Wextra -g -O2 -fPIC -I/usr/include/x86_64-linux-gnu -c -o src/avpacket_utils.o src/avpacket_utils.c
cc -o icecast-hls src/ini.o src/map.o src/main.o src/membuf.o src/miniflac.o src/minifmp4.o src/codecs.c src/decoder.o src/decoder_plugin.o src/decoder_plugin_auto.o src/decoder_plugin_miniflac.o src/decoder_plugin_passthrough.o src/demuxer.o src/demuxer_plugin.o src/demuxer_plugin_auto.o src/demuxer_plugin_flac.o src/demuxer_plugin_ogg.o src/destination.o src/destinationlist.o src/destination_sync.o src/encoder.o src/encoder_plugin.o src/encoder_plugin_passthrough.o src/filter.o src/filter_plugin.o src/filter_plugin_passthrough.o src/frame.o src/hls.o src/ich_time.o src/id3.o src/input.o src/input_plugin.o src/input_plugin_file.o src/input_plugin_stdin.o src/logger.o src/muxer.o src/muxer_plugin.o src/muxer_plugin_adts.o src/muxer_plugin_flac.o src/muxer_plugin_fmp4.o src/muxer_plugin_ogg.o src/muxer_plugin_ogg_flac.o src/muxer_plugin_ogg_opus.o src/muxer_plugin_packedaudio.o src/muxer_plugin_passthrough.o src/output.o src/output_plugin.o src/output_plugin_file.o src/output_plugin_icecast.o src/output_plugin_stdout.o src/output_plugin_folder.o src/packet.o src/samplefmt.o src/segment.o src/socket.o src/source.o src/sourcelist.o src/source_sync.o src/str.o src/strbuf.o src/tag.o src/tagmap.o src/tagmap_default.o src/thread.o src/version.o src/filter_plugin_avfilter.o src/demuxer_plugin_avformat.o src/encoder_plugin_avcodec.o src/decoder_plugin_avcodec.o src/encoder_plugin_opus.o src/avcodec_utils.o src/avframe_utils.o src/avpacket_utils.o  -L/usr/local/lib -lopus -lavformat -lavfilter -lavcodec -lavutil
/usr/bin/ld: src/thread.o: undefined reference to symbol 'pthread_getspecific@@GLIBC_2.2.5'
/usr/bin/ld: /lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make: *** [Makefile:275: icecast-hls] Error 1
```

OS info:
```
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.6 LTS
Release:	20.04
Codename:	focal
```

Thank you!

**Comment by triangle-stream** [comment]
It seems that I had two issues:

- ld was searching for libpthread**s**.so but I found that the file name was actually without the "S". To solve this I searched for the libpthread.so file and linked it to libpthreads.so, like this: `ln -s /usr/lib/x86_64-linux-gnu/libpthread.so /usr/local/lib/x86_64-linux-gnu/libpthreads.so`
- then, a new error was showed: `undefined reference to symbol 'pthread_create@@GLIBC_2.2.5](https://askubuntu.com/questions/376204/undefined-reference-to-symbol-pthread-createglibc-2-2-5-while-compiling-boc)`. Instead of editing the Makefile, I just added `LDFLAGS='-pthread'` to the make command.

**Comment by jprjr** [comment]
Oh right, that might be an issue with older GCC versions? I'm on Arch where I don't need the `-pthread` flag but older distros may need it.

I'll test on a few distros - it may need to be added to `CFLAGS` as well as `LDFLAGS`

**Comment by rlarzac** [comment]
Hello gentlemen, 
I am also facing compiling issues even if I remove Exhale. Would you have some hints to solve this issue ?

$ make ENABLE_EXHALE=0
cc -o icecast-hls src/ini.o src/map.o src/main.o src/membuf.o src/miniflac.o src/minifmp4.o src/codecs.o src/adts_mux.o src/ts.o src/decoder.o src/decoder_plugin.o src/decoder_plugin_auto.o src/decoder_plugin_miniflac.o src/decoder_plugin_passthrough.o src/demuxer.o src/demuxer_plugin.o src/demuxer_plugin_auto.o src/demuxer_plugin_flac.o src/demuxer_plugin_ogg.o src/destination.o src/destinationlist.o src/destination_sync.o src/encoder.o src/encoder_plugin.o src/encoder_plugin_passthrough.o src/filter.o src/filter_plugin.o src/filter_plugin_passthrough.o src/frame.o src/hls.o src/ich_time.o src/id3.o src/input.o src/input_plugin.o src/input_plugin_file.o src/input_plugin_stdin.o src/logger.o src/muxer.o src/muxer_plugin.o src/muxer_plugin_adts.o src/muxer_plugin_flac.o src/muxer_plugin_fmp4.o src/muxer_plugin_ogg.o src/muxer_plugin_ogg_flac.o src/muxer_plugin_ogg_opus.o src/muxer_plugin_packedaudio.o src/muxer_plugin_passthrough.o src/muxer_plugin_ts.o src/output.o src/output_plugin.o src/output_plugin_file.o src/output_plugin_icecast.o src/output_plugin_stdout.o src/output_plugin_folder.o src/packet.o src/samplefmt.o src/segment.o src/socket.o src/source.o src/sourcelist.o src/source_sync.o src/str.o src/strbuf.o src/tag.o src/tagmap.o src/tagmap_default.o src/thread.o src/version.o src/filter_plugin_avfilter.o src/demuxer_plugin_avformat.o src/encoder_plugin_avcodec.o src/decoder_plugin_avcodec.o src/encoder_plugin_fdk_aac.o src/input_plugin_curl.o src/output_plugin_curl.o src/encoder_plugin_opus.o src/avcodec_utils.o src/avframe_utils.o src/avpacket_utils.o  -lfdk-aac -lopus -lavformat -lavfilter -lavcodec -lavutil -lcurl
/usr/bin/ld : src/encoder_plugin.o:(.data.rel+0x10) : référence indéfinie vers « encoder_plugin_exhale »
collect2: error: ld returned 1 exit status
make: *** [Makefile:282 : icecast-hls] Erreur 1

Thank you for your help,
rlarzac

**Comment by jprjr** [comment]
Fixed in v1.2.0 13e9e4cd8e2b66199c52798a3ab276361c43a3c4
